shared_model/builders/protobuf/proto_transaction_builder.hpp

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_PROTO_TRANSACTION_BUILDER_HPP
+#define IROHA_PROTO_TRANSACTION_BUILDER_HPP
+
+#include "block.pb.h"
+#include "backend/protobuf/transaction.hpp"
+#include "interfaces/common_objects/types.hpp"
+#include "amount/amount.hpp"
+
+namespace shared_model {
+  namespace proto {
+    template <int S = 0>
+    class TemplateTransactionBuilder {
+     private:
+      template <int>
+      friend class TemplateTransactionBuilder;
+
+      enum RequiredFields { Command, CreatorAccountId, TxCounter, TOTAL };
+
+      template <int s>
+      using NextBuilder = TemplateTransactionBuilder<S | (1 << s)>;
+
+      iroha::protocol::Transaction transaction_;
+      iroha::protocol::Transaction::Payload &payload_;
+
+      template <int Sp>
+      operator TemplateTransactionBuilder<Sp>&() {
+        return reinterpret_cast<TemplateTransactionBuilder<Sp>&>(*this);
```

Comment:
Looks like a some hackish solution. May you explain that class and this operator especially?

Reply:
This is done for returning reference to another template type when a field is set, without redundant call of copy or move constructors. I will leave a comment in the code.